### ISSUE TITLE
Empty output when batch translating single character sentence

### ISSUE THREAD

**Issue by robertBrnnn** [OpenNMT/OpenNMT-py#2039]
Hi,

I have noticed an issue when batch translating content with my model, it appears to only occur for single character strings when translated in a batch. When the strings are translated in isolation there isn't an issue.
 
I've a sentence segmenter that segments text pre-translation, the resulting sentences are batch translated against the model. 

For instance, the text "1. An introduction to Machine Translation", would be split into the following two sentences "1. " and "An introduction to Machine Translation". 

I've instantiated a `onmt.translate.translator.Translator` and am sending the batch, as follows:
```
tokenized = [['▁', '1', '.', '▁'], ['▁An', '▁introduction', '▁to', '▁Machine', '▁Translation']]
scores, predictions = self.translator.translate(tokenized,
                                                batch_size=len(tokenized),
                                                batch_type='sents')
predictions = sum(predictions, [])
```
The returned predictions in this case are:
```
[ ' ||| ',   '▁机 器 翻译 简介 ||| 1-3 3-0 3-1 4-2']
```

The same phenomenon doesn't occur with full sentences, for instance: `tokenized = [['▁Next', '▁Chapter', '.', '▁'], ['▁An', '▁introduction', '▁to', '▁Machine', '▁Translation']]`, the returned translation is `['▁下 一 章 ||| 0-0 1-1 1-2', '▁机 器 翻译 简介 ||| 1-3 3-0 3-1 4-2']`.

As I said above the issue only occurs when batch translating these short strings, for instance translating `tokenized = [['▁', '1', '.', '▁']]` i.e. only one string, the returned translation is `['▁ 1 . ▁ ||| 0-0 1-1 2-2 3-3']`. There will only be an empty output, if there is more than one string in the batch, which is quite odd.

I'm using OpenNMT-py 2.0.1.

Any ideas what might be causing this? 

Cheers,
Rob

**Comment by francoishernandez** [comment]
This might be specific to your model and the data it was trained on. Technically the output sequence is not empty, since it has your separator token `'|||'`.

**Comment by robertBrnnn** [comment]
Hi @francoishernandez 

If this was specific to my model, shouldn't the output be the same regardless of whether the string is translated on it's own or in a batch? 

Maybe my initial post wasn't clear.

The below is current behavior:
```
# Two strings in the batch
tokenized = [['▁', '1', '.', '▁'], ['▁An', '▁introduction', '▁to', '▁Machine', '▁Translation']]
scores, predictions = self.translator.translate(tokenized,
                                                batch_size=len(tokenized),
                                                batch_type='sents')

# No translation for "1. "
predictions = sum(predictions, [])  # [ ' ||| ',   '▁机 器 翻译 简介 ||| 1-3 3-0 3-1 4-2']

# Single string in the batch, with the string the wasn't translated above
tokenized = [['▁', '1', '.', '▁']]
scores, predictions = self.translator.translate(tokenized,
                                                batch_size=len(tokenized),
                                                batch_type='sents')

# Translation returned for "1. "
predictions = sum(predictions, [])  # ['▁ 1 . ▁ ||| 0-0 1-1 2-2 3-3']
```

I'm getting different results depending on whether I translate a batch of multiple strings versus a batch with one string. 

Apologies if I'm missing something, but I expected the respective translations would be consistent regardless of whether there was one string in a batch, versus 2 or 5 strings in the batch.

**Comment by francoishernandez** [comment]
Oh, right, I read that a bit too quick.
Probably linked to #1320 then. You might want to play with the various `length_penalty` options.

**Comment by robertBrnnn** [comment]
No worries!

The length penalty and coverage penalty are currently set to wu, with a value of 0.2 for each. I'll try it out with length penalty as avg and tweak the alpha value to see if that helps.

**Comment by guillaumekln** [comment]
> Apologies if I'm missing something, but I expected the respective translations would be consistent regardless of whether there was one string in a batch, versus 2 or 5 strings in the batch.

In theory yes, but in practice batch computation has slightly different numerical results. This is a known side-effect of floating point computation. I think this is what you are observing.

When running on CPU there is a way to fix these differences. PyTorch is using Intel MKL which provides [a mode to make floating point computation reproducible](https://software.intel.com/content/www/us/en/develop/articles/introduction-to-the-conditional-numerical-reproducibility-cnr.html). In that case you can try to set the following environment variable and you should observe the same output (on CPU only):

```
MKL_CBWR=AUTO,STRICT
```

**Comment by robertBrnnn** [comment]
Thanks for the insight Guillaume, I wasn't aware of this side effect, good to know. 
I'm running these on GPU, I haven't noticed a similar issue on CPU, for which I'm using CTranslate2.

Interestingly, it seems if I disable the coverage penalty altogether, and leave the defaults I set for length penalty of wu and alpha 0.2, the issue doesn't occur and output is consistent. 

If coverage penalty is enabled, regardless of the beta values I tried, the output was inconsistent.
